Mathlib/Algebra/PresentedMonoid/Basic.lean

```diff
@@ -0,0 +1,141 @@
+/-
+Copyright (c) 2024. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Hannah Fechtner
+-/
+
+import Mathlib.Algebra.FreeMonoid.Basic
+import Mathlib.Algebra.Group.Submonoid.Operations
+import Mathlib.GroupTheory.Congruence.Hom
+
+/-!
+# Defining a monoid given by generators and relations
+
+Given relations `rels` on the free monoid on a type `α`, this file constructs the monoid
+given by generators `x : α` and relations `rels`.
+
+## Main definitions
+
+* `PresentedMonoid rels`: the quotient of the free monoid on a type `α` by the closure of one-step
+  reductions (arising from a binary relation on free monoid elements `rels`).
+* `PresentedMonoid.of`: The canonical map from `α` to a presented monoid with generators `α`.
+* `PresentedMonoid.lift f`: the canonical monoid homomorphism `PresentedMonoid rels → M`, given
+  a function `f : α → G` from a type `α` to a monoid `M` which satisfies the relations `rels`.
+
+## Tags
+
+generators, relations, monoid presentations
+-/
+
+variable {α : Type*}
+
+/-- Given a set of relations, `rels`, over a type `α`, `PresentedMonoid` constructs the monoid with
+generators `x : α` and relations `rels` as a quotient of a congruence structure over rels. -/
+@[to_additive "Given a set of relations, `rels`, over a type `α`, `PresentedAddMonoid` constructs
+the monoid with generators `x : α` and relations `rels` as a quotient of an AddCon structure over
+rels"]
+def PresentedMonoid (rel : FreeMonoid α → FreeMonoid α → Prop) := (conGen rel).Quotient
+
+namespace PresentedMonoid
+
+open Set Submonoid
+
+
+@[to_additive]
+instance {rels : FreeMonoid α → FreeMonoid α → Prop} : Monoid (PresentedMonoid rels) :=
+  Con.monoid (conGen rels)
+
+/-- The quotient map from the free monoid on `α` to the presented monoid with the same generators
+and the given relations `rels`. -/
+@[to_additive "The quotient map from the free additive monoid on `α` to the presented additive
+monoid with the same generators and the given relations `rels`"]
+def mk (rels : FreeMonoid α → FreeMonoid α → Prop) : FreeMonoid α →* PresentedMonoid rels where
```

Comment:
I think `surjective_mk` would be useful too.

Reply:
added! it's a bit below, because the proof uses PresentedMonoid.inductionOn